src/compute/src/logging/differential.rs

```diff
@@ -49,12 +49,13 @@ pub fn construct<A: Allocate>(
     let granularity_ms = std::cmp::max(1, config.granularity_ns / 1_000_000) as Timestamp;
 
     let traces = worker.dataflow_named("Dataflow: differential logging", move |scope| {
-        let logs = Some(linked).mz_replay(
+        let (logs, token) = Some(linked).mz_replay(
             scope,
             "differential logs",
             Duration::from_nanos(config.granularity_ns as u64),
             activator,
         );
+        std::mem::forget(token);
```

Comment:
Should we point out that there is a potential memory leak? I think it can show up when dropping the instance, lingering, and creating a new one on the same process.

Reply:
Yeah, I couldn't tell whether it was enough of a worry to thread a new argument through `mz_replay` to say "I don't even want a token" or to push the token through to wherever we stash the arrangements for the logging. I'm flexible on that; maybe taking an argument to not mint a token is best?

Reply:
I think there is a way to hold on to the token like this:
```diff
diff --git a/src/compute/src/compute_state.rs b/src/compute/src/compute_state.rs
index ce8869604..a612137b8 100644
--- a/src/compute/src/compute_state.rs
+++ b/src/compute/src/compute_state.rs
@@ -434,11 +434,11 @@ impl<'a, A: Allocate, B: ComputeReplay> ActiveComputeState<'a, A, B> {
                 .insert(id, Antichain::from_elem(0));
             logger.log(ComputeEvent::Frontier(id, 0, 1));
         }
-        for (log, trace) in d_traces {
+        for (log, (trace, token)) in d_traces {
             let id = logging.active_logs[&log];
             self.compute_state
                 .traces
-                .set(id, TraceBundle::new(trace, errs.clone()));
+                .set(id, TraceBundle::new(trace, errs.clone()).with_drop(token));
             self.compute_state
                 .reported_frontiers
                 .insert(id, Antichain::from_elem(0));
diff --git a/src/compute/src/logging/differential.rs b/src/compute/src/logging/differential.rs
index 6cc2e5033..e40ec7927 100644
--- a/src/compute/src/logging/differential.rs
+++ b/src/compute/src/logging/differential.rs
@@ -9,7 +9,9 @@

 //! Logging dataflows for events generated by differential dataflow.

+use std::any::Any;
 use std::collections::HashMap;
+use std::rc::Rc;
 use std::time::Duration;

 use differential_dataflow::collection::AsCollection;
@@ -43,9 +45,9 @@ use mz_timely_util::replay::MzReplay;
 pub fn construct<A: Allocate>(
     worker: &mut timely::worker::Worker<A>,
     config: &mz_dataflow_types::logging::LoggingConfig,
-    linked: std::rc::Rc<EventLink<Timestamp, (Duration, WorkerIdentifier, DifferentialEvent)>>,
+    linked: Rc<EventLink<Timestamp, (Duration, WorkerIdentifier, DifferentialEvent)>>,
     activator: RcActivator,
-) -> HashMap<LogVariant, KeysValsHandle> {
+) -> HashMap<LogVariant, (KeysValsHandle, Option<Rc<dyn Any>>)> {
     let granularity_ms = std::cmp::max(1, config.granularity_ns / 1_000_000) as Timestamp;

     let traces = worker.dataflow_named("Dataflow: differential logging", move |scope| {
@@ -55,7 +57,8 @@ pub fn construct<A: Allocate>(
             Duration::from_nanos(config.granularity_ns as u64),
             activator,
         );
-        std::mem::forget(token);
+
+        let mut token = Some(token);

         let mut demux =
             OperatorBuilder::new("Differential Logging Demux".to_string(), scope.clone());
@@ -197,7 +200,7 @@ pub fn construct<A: Allocate>(
                     })
                     .arrange_named::<RowSpine<_, _, _, _>>(&format!("ArrangeByKey {:?}", variant))
                     .trace;
-                result.insert(variant, trace);
+                result.insert(variant, (trace, token.take()));
             }
         }
         result
```

I can change your PR to capture this, or do it later.

Reply:
Ah, I can do that. I just finished typing a more complicated version than this, but it seems easier your way.

Reply:
I did the above, except with `token.clone()` rather than `token.take()`, to ensure that each arrangement keeps the dataflow alive (idk in the future maybe we uninstall some of the traces not all of them).

Reply:
Looks great, I'm more comfortable with this than `forget`ting the token.